### ISSUE TITLE
perf bug[venom]: unlifted variables in mem2var

### ISSUE THREAD

**Issue by charles-cooper** [vyperlang/vyper#4296]
### Version Information

* vyper Version (output of `vyper --version`): 6606ded787184be82de9ea05aeb1b61bcd219d6a

### What's your issue about?

examples of unlifted variables:
- loop variable
```
@external
def loop(x: uint256) -> uint256:
    res: uint256 = 9
    for i: uint256 in range(res, bound=10):
        res = res + i
    return res
```
![Screenshot from 2024-10-13 17-05-40](https://github.com/user-attachments/assets/2c3bdbe1-4b70-41e7-adc3-aedbda7c4a41)

- unlifted assignment
```
@external
def foo(x: uint256) -> uint256:
    t: uint256 = 5 // x
    return 6 // x
```
![image](https://github.com/user-attachments/assets/e883ea63-4c62-4da8-b2ed-872ba05b1937)

### How can it be fixed?

Fill this in if you know how to fix it.

**Comment by harkal** [comment]
Looks like the "transpiler" is not generating the appropriate `allocas`

**Comment by charles-cooper** [comment]
hmm. we should probably make sure in the transpiler that allocas get inserted in the entry block.

**Comment by harkal** [comment]
Looked into it. The `alloca` is fine. We need to do memory analysis to lift those `mstores`.

**Comment by charles-cooper** [comment]
i think the check for `all(use.opcode == "mstore") for use in uses` check in mem2var is blocking the second example:
https://github.com/vyperlang/vyper/blob/6843e7915729f3a3ea0d8c765dffa52033f5818e/vyper/venom/passes/mem2var.py#L43-L44

**Comment by charles-cooper** [comment]
this is blocking the first example, since the venom code contains a store of the alloca variable (`%ret_ofst = %11`):

https://github.com/vyperlang/vyper/blob/6843e7915729f3a3ea0d8c765dffa52033f5818e/vyper/venom/passes/mem2var.py#L45

**Comment by harkal** [comment]
Right. But it is not safe to lift those atm, as it fails. If alloca's were super correct the "all mstores" should have been a problem, but it probably not :/

**Comment by charles-cooper** [comment]
> Right. But it is not safe to lift those atm, as it fails. If alloca's were super correct the "all mstores" should have not been a problem, but it probably not correct :/

my diagnosis is:
- the first two guards in mem2var protect the calling convention, they are overly conservative. we should be able to add analysis to check which mloads/mstores are part of the calling convention
- ret_ofst should probably not be magical, let me look into this more
